doc/developer_guide/developer_guide.rst

```diff
@@ -0,0 +1,341 @@
+- Encapsulate running the tests and all its dependencies into docker
+  containers or volumes.
+- Error resilient
+
+Programming Model
+-----------------
+
+Exaslct is a mix of a build system, test runner and infrastructure as
+code. As such, we typically have tasks like the following one:
+
+- Build Image
+- Start Container
+- Upload something
+
+Most of these tasks produce some kind of output, for example: \* docker
+image \* a running docker container
+
+Often, other tasks then depend either on the output, or the action of
+one or more other tasks. These dependencies build a direct acyclic graph
+of tasks, also known as workflow.
+
+Tasks that depend on each other need to run in sequence, but tasks which
+are independent of each other may run in parallel. This model also
+allows a good separation of concern, because each Task solves one
+Problem.
+
+As workflow executor, we use
+`Luigi <https://luigi.readthedocs.io/en/stable/>`__ which was actually
+developed for batch data science workflows, but is suitable for other
+scenarios, too. Luigi describes tasks as subclasses of Luigi.Task which
+implements the following methods:
+
+::
+
+   class TaskC(luigi.Task):
+
+       def output(self):
+           return Target()
+
+       def run(self):
+           #do somthing
+           pass
+
+       def requires(self):
+           return [TaskA(),TaskB()]
+
+Here we describe a TaskC which depends on TaskA and TaskB defined in the
+``requires()`` method. It does something which is specified in the run()
+method. Further, it produces Target() as output. Luigi provides the
+dependency resolution, scheduling and parallelization.
+
+Besides, this static way of describing the dependencies between tasks,
+Luigi also provides so called `dynamic
+dependencies <https://luigi.readthedocs.io/en/stable/tasks.html#dynamic-dependencies>`__,
+which allow more flexible patterns in special case. Especially, if the
+order of execution of dependencies is important, or the dependencies
+depend on some calculation. The dynamic dependencies allow the
+implementation of a fork-join pattern.
+
+In EXASLCT we use our own subclass of Luigi.Task, ``StoppableTask`` as
+base class. The StoppableTask adds profiling, recording of dependencies
+for visualization or debugging and stops if any other StoppableTask
+failed in the workflow.
+
+Build Steps and Their Dependencies
+----------------------------------
+
+We compose the language container from several Dockerfiles. Each
+Dockerfile installs dependencies for one specific purpose. We also added
+a separate Dockerfile flavor-customization for user specific changes.
+The user specific changes will be merged on filesystem basis with the
+resulting docker images for the script client. The merge will overwrite
+user specific changes that could prevent the script client from working
+properly.
+
+The following graph shows the default build steps and their
+dependencies.
+
+.. image:: ./images/image-dependencies.png
+  :width: 400
```

Comment:
could make width a bit bigger. definitely out of scope, but diagram not style friendly (dark mode hides labels).